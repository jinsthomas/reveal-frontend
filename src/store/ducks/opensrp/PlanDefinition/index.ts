import { Dictionary } from '@onaio/utils/dist/types/types';
import intersect from 'fast_array_intersect';
import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import { createSelector } from 'reselect';
import SeamlessImmutable from 'seamless-immutable';
import { PlanDefinition } from '../../../../configs/settings';
import { isPlanDefinitionOfType } from '../../../../helpers/utils';
import { InterventionType } from '../../plans';

/** the reducer name */
export const reducerName = 'PlanDefinition';

// actions

/** PLAN_DEFINITION_FETCHED action type */
export const PLAN_DEFINITIONS_FETCHED =
  'reveal/reducer/opensrp/PlanDefinition/PLAN_DEFINITIONS_FETCHED';

/** PLAN_DEFINITION_FETCHED action type */
export const REMOVE_PLAN_DEFINITIONS =
  'reveal/reducer/opensrp/PlanDefinition/REMOVE_PLAN_DEFINITIONS';

/** PLAN_DEFINITION_FETCHED action type */
export const ADD_PLAN_DEFINITION = 'reveal/reducer/opensrp/PlanDefinition/ADD_PLAN_DEFINITION';

/** interface for fetch PlanDefinitions action */
interface FetchPlanDefinitionsAction extends AnyAction {
  planDefinitionsById: { [key: string]: PlanDefinition };
  planIdsByUserName: { [key: string]: string[] };
  type: typeof PLAN_DEFINITIONS_FETCHED;
}

/** interface for removing PlanDefinitions action */
interface RemovePlanDefinitionsAction extends AnyAction {
  planDefinitionsById: { [key: string]: PlanDefinition };
  planIdsByUserName: Dictionary<string[]>;
  type: typeof REMOVE_PLAN_DEFINITIONS;
}

/** interface for adding a single PlanDefinitions action */
interface AddPlanDefinitionAction extends AnyAction {
  planObj: PlanDefinition;
  type: typeof ADD_PLAN_DEFINITION;
}

/** Create type for PlanDefinition reducer actions */
export type PlanDefinitionActionTypes =
  | FetchPlanDefinitionsAction
  | RemovePlanDefinitionsAction
  | AddPlanDefinitionAction
  | AnyAction;

// action creators

/**
 * Fetch Plan Definitions action creator
 * @param {PlanDefinition[]} planList - list of plan definition objects
 * @param {string[]} userNames - openMRS user who has access to this plans.
 */
export const fetchPlanDefinitions = (
  planList: PlanDefinition[] = [],
  userNames: string[] = []
): FetchPlanDefinitionsAction => {
  const planIdsByUserName: Dictionary<string[]> = {};
  if (userNames.length > 0) {
    const plansIds = planList.map(plan => plan.identifier);
    userNames.forEach(userName => (planIdsByUserName[userName] = plansIds));
  }
  return {
    planDefinitionsById: keyBy(planList, 'identifier'),
    planIdsByUserName,
    type: PLAN_DEFINITIONS_FETCHED,
  };
};

/** Reset plan definitions state action creator */
export const removePlanDefinitions = () => ({
  planDefinitionsById: {},
  planIdsByUserName: {},
  type: REMOVE_PLAN_DEFINITIONS,
});

/**
 * Add one Plan Definition action creator
 * @param {PlanDefinition} planObj - the plan definition object
 */
export const addPlanDefinition = (planObj: PlanDefinition): AddPlanDefinitionAction => ({
  planObj,
  type: ADD_PLAN_DEFINITION,
});

// the reducer

/** interface for PlanDefinition state */
interface PlanDefinitionState {
  planDefinitionsById: { [key: string]: PlanDefinition } | {};
  planIdsByUserName: Dictionary<string[]> | {};
}

/** immutable PlanDefinition state */
export type ImmutablePlanDefinitionState = PlanDefinitionState &
  SeamlessImmutable.ImmutableObject<PlanDefinitionState>;

/** initial PlanDefinition state */
const initialState: ImmutablePlanDefinitionState = SeamlessImmutable({
  planDefinitionsById: {},
  planIdsByUserName: {},
});

/** the PlanDefinition reducer function */
export default function reducer(
  state = initialState,
  action: PlanDefinitionActionTypes
): ImmutablePlanDefinitionState {
  switch (action.type) {
    case PLAN_DEFINITIONS_FETCHED:
      if (action.planDefinitionsById) {
        return SeamlessImmutable({
          ...state,
          planDefinitionsById: { ...state.planDefinitionsById, ...action.planDefinitionsById },
          planIdsByUserName: { ...state.planIdsByUserName, ...action.planIdsByUserName },
        });
      }
      return state;
    case ADD_PLAN_DEFINITION:
      if (action.planObj as PlanDefinition) {
        return SeamlessImmutable({
          ...state,
          planDefinitionsById: {
            ...state.planDefinitionsById,
            [action.planObj.identifier as string]: action.planObj,
          },
        });
      }
      return state;
    case REMOVE_PLAN_DEFINITIONS:
      return SeamlessImmutable({
        ...state,
        planDefinitionsById: action.planDefinitionsById,
        planIdsByUserName: action.planIdsByUserName,
      });
    default:
      return state;
  }
}

// selectors

/** get PlanDefinitions by id
 * @param {Partial<Store>} state - the redux store
 * @param {InterventionType | null} interventionType - the PlanDefinition intervention Type
 * @returns {{ [key: string]: PlanDefinition }} PlanDefinitions by id
 */
export function getPlanDefinitionsById(
  state: Partial<Store>,
  interventionType: InterventionType | null = null
): { [key: string]: PlanDefinition } {
  if (interventionType) {
    return keyBy(getPlanDefinitionsArray(state, interventionType), 'identifier');
  }
  return (state as any)[reducerName].planDefinitionsById;
}

/** get one PlanDefinition using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} id - the PlanDefinition id
 * @returns {PlanDefinition|null} a PlanDefinition object or null
 */
export function getPlanDefinitionById(state: Partial<Store>, id: string): PlanDefinition | null {
  return get((state as any)[reducerName].planDefinitionsById, id) || null;
}

/** get an array of PlanDefinition objects
 * @param {Partial<Store>} state - the redux store
 * @param {InterventionType | null} interventionType - the PlanDefinition intervention Type
 * @returns {PlanDefinition[]} an array of PlanDefinition objects
 */
export function getPlanDefinitionsArray(
  state: Partial<Store>,
  interventionType: InterventionType | null = null
): PlanDefinition[] {
  const result = values((state as any)[reducerName].planDefinitionsById);
  if (interventionType) {
    return result.filter((e: PlanDefinition) => isPlanDefinitionOfType(e, interventionType));
  }
  return result;
}

/** RESELECT USAGE STARTS HERE */

/** This interface represents the structure of plan definition filter options/params */
export interface PlanDefinitionFilters {
  title?: string /** plan object title */;
  userName?: string /** openMRS user, filter out plans that the user does not have access to */;
}

/** planDefinitionsArrayBaseSelector select an array of all plans
 * @param state - the redux store
 */
export const planDefinitionsArrayBaseSelector = (planKey = 'planDefinitionsById') => (
  state: Partial<Store>
) => values((state as any)[reducerName][planKey]);

/** getTitle
 * Gets title from PlanFilters
 * @param state - the redux store
 * @param props - the plan filters object
 */
export const getTitle = (_: Partial<Store>, props: PlanDefinitionFilters) => props.title;

/** non_memoized function that gets value of reducerSlice.planDefinitionsById
 * @param {Partial<Store>} state -  the redux store
 */
export const basePlanDefinitionsByIds = (state: Partial<Store>): Dictionary<PlanDefinition> =>
  (state as any)[reducerName].planDefinitionsById;

/** non_memoized function that gets value of reducerSlice.planIdsByUserName
 * * @param {Partial<Store>} state -  the redux store
 */
export const basePlanIdsByUserNamesSelector = (state: Partial<Store>): Dictionary<string[]> => {
  return (state as any)[reducerName].planIdsByUserName;
};

/** getUserName
 * Gets userName from PlanFilters
 * @param state - the redux store
 * @param props - the plan filters object
 */
export const getUserName = (_: Partial<Store>, props: PlanDefinitionFilters) => props.userName;

/** gets all planIds associated with a userName
 * @param {Registry} state - the redux store
 * @param {PlanDefinitionFilters} props - the plan definition filters object
 */
export const getPlanIdsByUserName = () => {
  return createSelector(
    basePlanIdsByUserNamesSelector,
    getUserName,
    (planIdsByUserName, userName) => {
      let planIdsArray = userName ? planIdsByUserName[userName] : [];
      planIdsArray = planIdsArray ? planIdsArray : [];
      return planIdsArray;
    }
  );
};

/** filters out plans that the given user does not have access to.
 * @param {Registry} state - the redux store
 * @param {PlanDefinitionFilters} props - the plan definition filters object
 */
export const getPlansByUserName = () => {
  return createSelector(
    getPlanIdsByUserName(),
    getUserName,
    basePlanDefinitionsByIds,
    (planIds, userName, allPlansDictionary) => {
      const plansOfInterest = userName
        ? planIds.map((planId: string) => allPlansDictionary[planId])
        : values(allPlansDictionary);
      return plansOfInterest;
    }
  );
};

/** getPlansArrayByTitle
 * Gets an array of Plan objects filtered by plan title
 * @param {Registry} state - the redux store
 * @param {PlanDefinitionFilters} props - the plan definition filters object
 */
export const getPlanDefinitionsArrayByTitle = (planKey?: string) =>
  createSelector([planDefinitionsArrayBaseSelector(planKey), getTitle], (plans, title) =>
    title ? plans.filter(plan => plan.title.toLowerCase().includes(title.toLowerCase())) : plans
  );

/** makePlanDefinitionsArraySelector
 * Returns a selector that gets an array of IRSPlan objects filtered by one or all
 * of the following:
 *    - title
 *
 * These filter params are all optional and are supplied via the prop parameter.
 *
 * This selector is meant to be a memoized replacement for getPlanDefinitionsArray.
 *
 * To use this selector, do something like:
 *    const PlanDefinitionsArraySelector = makeIRSPlansArraySelector();
 *
 * @param {Partial<Store>} state - the redux store
 * @param {PlanFilters} props - the plan filters object
 * @param {string} sortField - sort by field
 */
export const makePlanDefinitionsArraySelector = (planKey?: string) => {
  return createSelector(
    [getPlanDefinitionsArrayByTitle(planKey), getPlansByUserName()],
    (plans1, plans2) => {
      return intersect([plans1, plans2], JSON.stringify);
    }
  );
};
