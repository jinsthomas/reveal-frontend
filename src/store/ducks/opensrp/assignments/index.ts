/** Assignments redux module */
import { get } from 'lodash';
import { Store } from 'redux';
import { AnyAction } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'assignments';

/** interface for a Assignment object */
export interface Assignment {
  jurisdiction: string;
  organization: string;
  plan: string;
  fromDate: string;
  toDate: string;
}

// action interfaces

/** action type for action that adds Assignments to store */
export const ASSIGNMENTS_FETCHED = 'src/store/ducks/assignments/reducer/ASSIGNMENTS_FETCHED';
/** action type for REMOVE_TEAMS action */
export const REMOVE_ASSIGNMENTS = 'src/store/ducks/assignments/reducer/REMOVE_ASSIGNMENTS';

/** interface for Assignments fetched action */
interface FetchAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof ASSIGNMENTS_FETCHED;
}

/** interface for action that removes assignments from store */
interface RemoveAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof REMOVE_ASSIGNMENTS;
}

/** single type for all action types */
type AssignmentActionTypes = FetchAssignmentsAction | RemoveAssignmentsAction | AnyAction;

/** interface for Assignments state in store */
interface AssignmentsStoreState {
  assignmentsByPlanId: { [key: string]: Assignment[] };
}

// immutable assignments state in dux
export type ImmutableAssignmentsStoreState = AssignmentsStoreState &
  SeamlessImmutable.ImmutableObject<AssignmentsStoreState>;

/** initial state for Assignments records in store */
const initialOrgsStoreState: ImmutableAssignmentsStoreState = SeamlessImmutable({
  assignmentsByPlanId: {},
});

/** the Assignment reducer function */
export default function reducer(state = initialOrgsStoreState, action: AssignmentActionTypes) {
  switch (action.type) {
    case ASSIGNMENTS_FETCHED:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: {
          ...state.assignmentsByPlanId,
          ...action.assignmentsByPlanId,
        },
      });
    case REMOVE_ASSIGNMENTS:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: action.assignmentsByPlanId,
      });
    default:
      return state;
  }
}

/** action to remove assignments form store */
export const removeAssignmentsAction: RemoveAssignmentsAction = {
  assignmentsByPlanId: {},
  type: REMOVE_ASSIGNMENTS,
};

// action creators

/** creates action to add fetched assignments to store
 * @param {Assignment []} assignmentsList - array of assignments to be added to store
 *
 * @returns {FetchAssignmentsAction} - action with assignments payload that is added to store
 */
export const FetchAssignments = (assignmentsList: Assignment[]): FetchAssignmentsAction => {
  const assignmentsByPlanId: { [key: string]: Assignment[] } = {};

  for (const assignment of assignmentsList) {
    if (!assignmentsByPlanId[assignment.plan]) {
      assignmentsByPlanId[assignment.plan] = [];
    }
    assignmentsByPlanId[assignment.plan].push(assignment);
  }

  return {
    assignmentsByPlanId,
    type: ASSIGNMENTS_FETCHED,
  };
};

// selectors

/** get assignments as an object where their ids are the keys and the objects
 * the values
 * @param {Partial<Store>} state - Portion of the store
 *
 * @return {[key: string]: Assignment}
 */
export function getAssignmentsByPlanId(state: Partial<Store>): { [key: string]: Assignment[] } {
  return (state as any)[reducerName].assignmentsByPlanId;
}

/** Get all assignments as an array
 * @param {Partial<Store>} state - Part of the redux store
 * @param {string} planId - The id of the plan of which to return associated Assignments
 *
 * @return {Assignment []} - all assignments by planId in store as an array
 */
export function getAssignmentsArrayByPlanId(state: Partial<Store>, planId: string): Assignment[] {
  return get(getAssignmentsByPlanId(state), planId);
}
