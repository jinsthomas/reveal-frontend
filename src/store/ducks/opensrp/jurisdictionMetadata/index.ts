/** interface for JurisdictionMetadataImport object */
export interface JurisdictionMetadataImport {
  file: File;
}
