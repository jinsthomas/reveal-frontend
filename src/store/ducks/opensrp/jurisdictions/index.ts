import { Dictionary } from '@onaio/utils/dist/types/types';
import {
  fetchActionCreatorFactory,
  reducerFactory,
  removeActionCreatorFactory,
} from '@opensrp/reducer-factory';
import { Feature, FeatureCollection, Geometry } from '@turf/turf';
import { get, values } from 'lodash';
import { Store } from 'redux';
import { createSelector } from 'reselect';

/** The shape of a jurisdiction received from the OpenSRP API */
export interface Jurisdiction {
  id: string;
  geometry?: Geometry;
  properties: {
    code?: string;
    geographicLevel: number;
    name: string;
    parentId?: string;
    status: string;
    version: string | number;
  };
  serverVersion: number;
  type: Readonly<'Feature'>;
}

/** the reducer name */
export const reducerName = 'Jurisdiction';

/** Jurisdiction Reducer */
const reducer = reducerFactory<Jurisdiction>(reducerName);
export default reducer;

// actions
/** actionCreator returns action to to add Item records to store */
export const fetchJurisdictions = fetchActionCreatorFactory<Jurisdiction>(reducerName, 'id');
export const removeJurisdictions = removeActionCreatorFactory(reducerName);

// selectors
/** prop filters to customize selector queries */
export interface Filters {
  jurisdictionId?: string /** jurisdiction id */;
  jurisdictionIdsArray?: string[] /** array of jurisdiction ids */;
  parentId?: string /** parent id */;
}

/** retrieve the jurisdictionId value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionId = (_: Partial<Store>, props: Filters) => props.jurisdictionId;

/** retrieve the parentId value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getParentId = (_: Partial<Store>, props: Filters) => props.parentId;

/** retrieve the jurisdictionIdsArray value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionIdsArray = (_: Partial<Store>, props: Filters) =>
  props.jurisdictionIdsArray;

/** gets all jurisdictions keyed by id
 * @param state - the store
 * @param _ -  the filterProps
 */
export const getJurisdictionsById = (state: Partial<Store>, _: Filters): Dictionary<Jurisdiction> =>
  (state as any)[reducerName].objectsById;

/** retrieve the Jurisdiction using an id
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionById = () =>
  createSelector(
    [getJurisdictionsById, getJurisdictionId],
    (jurisdictionsById, jurisdictionId): Jurisdiction | null => {
      if (!jurisdictionId) {
        return null;
      }
      return get(jurisdictionsById, jurisdictionId, null);
    }
  );

/**
 * Get jurisdiction ids
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionIds = () =>
  createSelector([getJurisdictionsById], (jurisdictionsById): string[] => {
    return Object.keys(jurisdictionsById);
  });

/**
 * retrieve jurisdictions as an array
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionsArray = () =>
  createSelector(
    [getJurisdictionsById, getParentId, getJurisdictionIdsArray],
    (jurisdictionsById, parentId, jurisdictionIdsArray): Jurisdiction[] => {
      const result = values(jurisdictionsById);
      if (parentId) {
        return result.filter(item => item.properties.parentId === parentId);
      }
      if (jurisdictionIdsArray) {
        return result.filter(item => jurisdictionIdsArray.includes(item.id));
      }
      return result;
    }
  );

/**
 * retrieve jurisdictions as a geojson feature collection
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionsFC = () =>
  createSelector(
    [getJurisdictionsArray()],
    (jurisdictionsArray): FeatureCollection => {
      return {
        features: jurisdictionsArray.filter(item => 'geometry' in item) as Feature[],
        type: 'FeatureCollection',
      };
    }
  );
