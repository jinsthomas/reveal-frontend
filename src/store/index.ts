import { combine } from '@onaio/redux-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { applyMiddleware, compose, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import users, { reducerName as usersReducer } from './ducks/users';

/** Add redux devtools to Window */
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

/** Declare history variable */
export const history = createBrowserHistory();

/** Declare type for reducer Registry */
export interface Registry {
  [key: string]: Reducer;
}

/** Function to create the store */
export function getStore(reducers: Registry) {
  /** Register each of the initial reducers */
  Object.keys(reducers).forEach(reducerName => {
    reducerRegistry.register(reducerName, reducers[reducerName]);
  });

  /** Combine reducers */
  const reducer = combine(reducerRegistry.getReducers());

  /** Add redux devtools to enhancers */
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  /** Create the store */
  return createStore(reducer, composeEnhancers(applyMiddleware(thunk, routerMiddleware(history))));
}

/** Router reducer */
const connectReducer = connectRouter(history);

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[usersReducer] = users;

/** The initial store for the reveal web app */
const store = getStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
