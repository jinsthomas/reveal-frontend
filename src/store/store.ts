import { combine } from '@onaio/redux-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { applyMiddleware, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import users, { reducerName as usersReducer } from './ducks/users';

export const history = createBrowserHistory();

/** Declare type for reducer Registry */
export interface Registry {
  [key: string]: Reducer;
}

/** Function to create the store */
export function getStore(reducers: Registry) {
  /** Register each of the initial reducers */
  Object.keys(reducers).forEach(reducerName => {
    reducerRegistry.register(reducerName, reducers[reducerName]);
  });
  /** Combine reducers */
  const reducer = combine(reducerRegistry.getReducers());
  /** Create the store */
  return createStore(
    reducer,
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__(),
    applyMiddleware(thunk, routerMiddleware(history))
  );
}

/** Router reducer */
const connectReducer = connectRouter(history);
/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};
// /** Add users reducer to registry */
defaultReducers[usersReducer] = users;

/** The initial store for the reveal web app */
const store = getStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
