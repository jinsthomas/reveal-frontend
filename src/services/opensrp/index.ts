import { OpenSRPService as OpenSRPServiceWeb } from '@opensrp/server-service';
import { IncomingHttpHeaders } from 'http';
import { OPENSRP_API_BASE_URL } from '../../configs/env';
import store from '../../store';
import { getAccessToken } from '../../store/selectors';

/** allowed http methods */
type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';

/** get default HTTP headers for OpenSRP service */
export function getDefaultHeaders(
  accept: string = 'application/json',
  authorizationType: string = 'Bearer',
  contentType: string = 'application/json;charset=UTF-8'
): IncomingHttpHeaders {
  return {
    accept,
    authorization: `${authorizationType} ${getAccessToken(store.getState())}`,
    'content-type': contentType,
  };
}

/** interface to describe URL params object */
export interface URLParams {
  [key: string]: string | number | boolean;
}

/** converts filter params object to string
 * @param {URLParams} obj - the object representing filter params
 * @returns {string} filter params as a string
 */
export function getFilterParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}:${val}`)
    .join(',');
}

/** get payload for fetch
 * @param {AbortSignal} signal - signal object that allows you to communicate with a DOM request
 * @param {HTTPMethod} method - the HTTP method
 * @returns the payload
 */

export function getPayloadOptions(_: AbortSignal, method: HTTPMethod) {
  return {
    headers: getDefaultHeaders() as HeadersInit,
    method,
  };
}

/** The OpenSRP service class
 * Extends class OpenSRPService from OpenSRPService web
 * Sample usage:
 * -------------
 * const service = new OpenSRPService('the-endpoint');
 *
 * **To list all objects**: service.list()
 *
 * **To get one object**: service.read('the-object-identifier')
 *
 * **To create a new object**: service.create(theObject)
 *
 * **To update an object**: service.update(theObject)
 */
export class OpenSRPService extends OpenSRPServiceWeb {
  constructor(
    endpoint: string,
    baseURL: string = OPENSRP_API_BASE_URL,
    getPayload: typeof getPayloadOptions = getPayloadOptions
  ) {
    super(baseURL, endpoint, getPayload);
  }
  /** create method
   * Send a POST request to the general endpoint containing the new object data
   * Successful requests will result in a HTTP status 201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async create<T>(data: T, params: any = null, method: HTTPMethod = 'POST'): Promise<{}> {
    const url = OpenSRPService.getURL(this.generalURL, params);
    const payload = {
      ...this.getOptions(this.signal, method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok || response.status !== 201) {
      throw new Error(
        `OpenSRPService create on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }
}
