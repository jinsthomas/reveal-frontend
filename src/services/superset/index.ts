import superset from '@onaio/superset-connector';
import { clone } from 'lodash';
import { CONFIG } from '../../configs/superset';
import store from '../../store';
import { getApiToken } from '../../store/selectors';

/** middleware for fetching from Superset */
export const fetchMiddleware = (res: { [key: string]: any }) => {
  return res;
};

/** callback for fetching from Superset */
export const fetchCallback = (parsedResponse: Array<{ [key: string]: any }>) => {
  const sliceData = superset.processData(parsedResponse);
};

/** generic async function for fetching from superset */
const supersetFetch = async (
  sliceId: string,
  callback: typeof fetchCallback = fetchCallback,
  middleware: typeof fetchMiddleware = fetchMiddleware
) => {
  const config = clone(CONFIG);
  config.extraPath = sliceId;
  config.token = getApiToken(store.getState());
  superset.api.fetch(config, middleware).then(callback);
};

export default supersetFetch;
