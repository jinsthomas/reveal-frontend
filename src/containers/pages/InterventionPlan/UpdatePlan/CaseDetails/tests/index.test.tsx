import {mount, shallow} from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import ConnectedCaseDetails, { CaseDetails } from '..';
import { OpenSRPService } from '../../../../../../services/opensrp';
import store from '../../../../../../store';
/** Test sequence
 *  - renders without crashing.
 *  - renders correctly - checks initial render
 *      - collapsible is not collapsed.
 *  - collapsible works correctly:
 *      - parse html text for correct content
 *  - works correctly with store
 *  - makes correct api calls
 */

const fetch = require('jest-fetch-mock');

const props = {
  eventId: '',
  service: OpenSRPService,
  event: {},
};

describe('src/containers/pages/interventionPlan/updateplan/caseDetials', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(<CaseDetails {...props} />);
  });

  it('renders correctly', () => {
    const wrapper = mount(<CaseDetails {...props} />);
    // accordion is not collapsed
    expect(wrapper.text()).toMatchInlineSnapshot();
  });

  it('Accordion works correctly', () => {
    const wrapper = mount(<CaseDetails {...props} />);
    const cardHeader = wrapper.find('CardHeader');
    // simulate click on cardHeader(Case Details)
    cardHeader.simulate('click');
    wrapper.update();
    // The case details should now be visible
    expect(wrapper.text()).toMatchInlineSnapshot();
  });

  it('works correctly with store', () => {
    fetch.once(JSON.stringify({}));
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedCaseDetails {...props} />
      </Provider>
    );
    // should basically be able to render the correct output
    expect(wrapper.find(CaseDetails).text()).toMatchInlineSnapshot();
  });

  it('Makes the correct api calls', () => {
    fetch.once(JSON.stringify({}));
    const wrapper = mount(<CaseDetails {...props} />);
    expect(fetch).toHaveBeenCalledWith({});
  })
});
