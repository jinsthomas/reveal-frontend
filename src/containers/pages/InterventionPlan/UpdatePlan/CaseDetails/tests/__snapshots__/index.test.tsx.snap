// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/containers/pages/interventionPlan/updateplan/caseDetials renders correctly: Case Details 1`] = `"Case Number: 191316000004593200125140915723Case DetailsCase InformationAge: 13Case Class.: ADate of Diagnosis: 2020-01-25House Number: Investigation Date: 2020-01-24Case Notif. Date: 2020-01-25Patient Name: ชาวาดีSpecies: VSurname: ดารากมล Foci of ResidenceClassification: A1id: 88684506-605d-41db-b904-efbaf9795d2aName: แม่หิต (6305010804)"`;

exports[`src/containers/pages/interventionPlan/updateplan/caseDetials works well with store: Case Details 1`] = `"Case Number: 191316000004593200125140915723"`;
